enterprise/deferred-locks/src/main/java/org/neo4j/kernel/impl/locking/DeferringLockClient.java

```diff
@@ -0,0 +1,194 @@
+    {
+        removeLock( resourceType, resourceId, true );
+    }
+
+    void acquireDeferredLocks()
+    {
+        assertNotStopped();
+
+        long[] current = new long[10];
+        int cursor = 0;
+        Locks.ResourceType currentType = null;
+        boolean currentExclusive = false;
+        for ( LockUnit lockUnit : locks.keySet() )
+        {
+            // TODO perhaps also add a condition which sends batches over a certain size threshold
+            if ( currentType == null ||
+                 (currentType.typeId() != lockUnit.resourceType().typeId() ||
+                  currentExclusive != lockUnit.isExclusive()) )
+            {
+                // New type, i.e. flush the current array down to delegate in one call
+                flushLocks( current, cursor, currentType, currentExclusive );
+
+                cursor = 0;
+                currentType = lockUnit.resourceType();
+                currentExclusive = lockUnit.isExclusive();
+            }
+
+            // Queue into current batch
+            if ( cursor == current.length )
+            {
+                current = Arrays.copyOf( current, cursor * 2 );
+            }
+            current[cursor++] = lockUnit.resourceId();
+        }
+        flushLocks( current, cursor, currentType, currentExclusive );
+    }
+
+    private void flushLocks( long[] current, int cursor, Locks.ResourceType currentType, boolean exclusive )
+    {
+        if ( cursor > 0 )
+        {
+            long[] resourceIds = Arrays.copyOf( current, cursor );
+            if ( exclusive )
+            {
+                clientDelegate.acquireExclusive( currentType, resourceIds );
+            }
+            else
+            {
+                clientDelegate.acquireShared( currentType, resourceIds );
+            }
+        }
+    }
+
+    @Override
+    public void stop()
+    {
+        stopped = true;
+        clientDelegate.stop();
+    }
+
+    @Override
+    public void close()
+    {
+        stopped = true;
+        clientDelegate.close();
+    }
+
+    @Override
+    public int getLockSessionId()
+    {
+        return clientDelegate.getLockSessionId();
+    }
+
+    private void assertNotStopped()
+    {
+        if ( stopped )
+        {
+            throw new LockClientStoppedException( this );
+        }
+    }
+
+    private void addLock( Locks.ResourceType resourceType, long resourceId, boolean exclusive )
+    {
+        assertNotStopped();
+
+        LockUnit lockUnit = new LockUnit( resourceType, resourceId, exclusive );
+        MutableInt lockCount = locks.get( lockUnit );
+        if ( lockCount == null )
+        {
+            lockCount = new MutableInt();
+            locks.put( lockUnit, lockCount );
+        }
+        lockCount.increment();
+    }
+
+    private void removeLock( Locks.ResourceType resourceType, long resourceId, boolean exclusive )
+    {
+        assertNotStopped();
```

Comment:
`removeLock` is called from `releaseShared` and `releaseExclusive`, probably better to check here and not duplicate as well.